sourcegraph/cody
# feat: new /doc command with smart selection

feat: add smart selection and update doc prompt

This PR focuses on updating the `/doc` command, featuring `snart selection`.

- Add getActiveTextEditorSmartSelection method to editor interface for supporting "smart selection" on files that have lang server running as we rely on that to extract the ranges for each function / methods within a class
- Update prompt for the /doc command to specify docstring should not contain code but docstring only
- Introduce `CommandRunner` class that allows different services to run different commands async
- Add `mode` field to cody commands config 
- Add `insertMode` and `autoApply` fields to FixupTask that allows fixup to auto apply suggestions returned by Cody instead of displaying Code Lenses
- `/doc` command now uses fixup insert mode 

What is `smart selection`?

Smart selection removes the need to manually highlight code before running the /doc and /test commands. Instead, it will automatically detect and run the selected command for the function nearest to your cursor position. For /doc, the docstring will be added to the top of the function through fixup.

Why?

Based on user testing conducted by Megan, many first-time command users were unaware of the requirement to highlight code before running commands. We can create a smoother onboarding experience for new users with smart selection. Allowing our commands to work seamlessly without manual code selection will lower the learning curve and provide a more intuitive first-time user experience. 

However, it's important to note that smart selection should be treated as a fallback method, as it produces a selection based on guessing user intent. Manual selection truly reflects the user's real intent and should be used when possible. Moreover, smart selection may not work well for all languages, and depends on if users have language servers running in their editor. In cases where smart selection is unreliable, users should fall back to using the manual selection method to ensure accuracy.

## Follow-up

- [ ] improve /doc prompt
- [x] apply fixup automatically
- [x] indentation issues caused by fixup

## Test plan



Update 2 - latest

https://github.com/sourcegraph/cody/assets/68532117/a928612b-c792-43f5-9947-29bd52f2ea17

Update 1

https://github.com/sourcegraph/cody/assets/68532117/02953c54-efe0-41dc-acdb-45ffb15e6404

1. Start Cody from this branch
2. Run `/doc` command in a function without selecting any code (but put your cursor inside a function)
3. Run `/doc` command with code selected
4. Repeat step 2 and 3 with `/test` command
5. Run other commands, and use chat -- they should work as expected
